### ISSUE TITLE
Infinite loop when upgrading to 5.7.0

### ISSUE THREAD

**Issue by slaskis** [developit/preact#277]
My app just keeps re-rendering the first component when I upgraded to 5.7.0, while any previous versions (5.6.0 down to 5.1.0-beta at least) works fine.

My app uses redux dispatching and some custom transitions modeled after reacts transition groups but it seems like a pretty big regression since they work nicely in older versions.

**Comment by developit** [comment]
ack! Any chance you can pause it to see what is triggering the change?

**Comment by slaskis** [comment]
I'm investigating my transition component right now, trying to see if any of the changes between 5.6 and 5.7 might be related to it.

Right now the transition + animation components render every frame until done (like an opacity animation), and it worked nicely before but now it just keeps rendering. So there might be something in there...

**Comment by slaskis** [comment]
It not only renders but seem to trigger mount too

**Comment by developit** [comment]
that might explain how it's looping, since redux subscribes on mount.  Ok diagnosis time (these might seem random!):

- Is there a `<select>` dropdown in the tree?
- Are you bundling with webpack, or browserify?
- Are you including a Promise polyfill? (if so, which one?)
- Are any of your event handlers bound specified with `on` being uppercase? (eg: `ONload`)

I have a beta locally that I'd really like you to try out, it's quite possible the issue is already fixed but has not yet been released.  I just finished my work up last night but need to do some testing before an official 5.8 release, but I can send you a preview or push to master and you can install from there (`npm i developit/preact#master`).

I'm in the [preact gitter room](https://gitter.im/developit/preact) if the back-and-forth on here is too slow.

**Comment by slaskis** [comment]
Actually, it doesn't seem related to rendering or mounting at all? I've added logs to those but they only log once, but it keeps triggering actions. So it's probably in the redux middlewares somehow. 

So this is most likely a bug somewhere else that might be made visible by the preact update? Still odd since downgrading to 5.6.0 "fixes" it. 

The reason I'm trying to upgrade though is because I'm having some trouble with `componentDidMount` not always being triggered on safari ios for some reason.

**Comment by slaskis** [comment]
Interesting questions! 

- I am using babel polyfill Promises. 
- Bundling with browserify. 
- There are no `<select>`s anywhere in the app. 
- And any events are using `onEvent` (small first letter, camelcased the rest)

**Comment by developit** [comment]
Interesting - that's the first I've heard of a Safari-specific bug there.  Thanks for the answers, those help immensely.  I was basically going through the list of commits between 5.6 and 5.7 looking for what changed and how that might be in use in your app.  All your answers check out, though I'm curious if 5.6 still has the bug for you but is just slower to infinitely loop (which would be pretty bad!).

You can actually log out every render pretty easily, which might reveal the source of the looping:

```js
import { options } from 'preact';

options.syncComponentUpdates = false;  // always re-render asynchronously

options.debounceRendering = fn => {
  console.trace('render');
  setTimeout(fn, 100);  // throttle updates at 100ms
};
```

**Comment by slaskis** [comment]
ooh, nice! Will check that out.

I was about to try the master branch of preact but for some reason budo/browserify just complains with `Cannot find module 'preact'`

**Comment by developit** [comment]
hmm. That makes sense, since there is no dist. I can release it as a beta tag, will be a few minutes.

**Comment by slaskis** [comment]
Ok, the plot thickens...I added that `debounceRendering` callback and well, it only renders once!

So it must be the redux actions that are looping...

**Comment by developit** [comment]
Yep!  So this makes sense then:  @insin pointed out in #261 that because preact's internal debouncing function was called `setImmediate()`, it was actually getting _replaced_ when built with Browserify or Webpack (crazy!).  I renamed it to `defer()`, since the semantics are different from Node's `setImmediate` - defer is just used to collate renders into a single frame, but is still blocking (that's why it uses `Promise` for tight async instead of something like `requestAnimationFrame`).

So, to sum up: something about your redux actions is looping when they are invoked synchronously. Prior to 5.7, your Browserify build was actually modifying preact in a way that made it wait a few milliseconds before renders, and that was hiding the issue.

In terms of what to look for when you're trying to find the offending action - I'd look for something that is depending on intermediary state, or two async actions that if invoked between eachother would create a loop.

**Comment by slaskis** [comment]
Hmm, great explanation, I think I know exactly where to look, will let you know how that goes!

**Comment by developit** [comment]
On the plus side, your whole app should be quite a bit faster without that timers shim being injected (once the infinite loop is dealt with haha).

**Comment by slaskis** [comment]
Now I'm super confused, it now renders and calls lifecycle events in the loop again.

So right now it seems like the dispatch (which i do in a componentDidMount callback) is being triggered, which on success triggers the componentDidMount again...and around it goes.

**Comment by slaskis** [comment]
It never calls `componentWillUnmount` in between though, might be a clue?

**Comment by developit** [comment]
Is this with the `options.syncComponentUpdates=false` bit added?  Try it just with this:

```js
import { options } from 'preact';

options.debounceRendering = fn => {
  console.trace('render');
  setTimeout(fn, 100);  // throttle updates at 100ms
};
```

**Comment by slaskis** [comment]
Hmm. I think I'll have to take a break from staring at it now. It's just making me more confused, every change I do causes something else to behave unexpectedly. 

Really appreciate your help though, hopefully I'll be able to provide some better testing later.

**Comment by slaskis** [comment]
Ok, to make matters worse. I figured I could try it in Chrome Canary to see if dev tools would give me more information. There it works just perfectly..

**Comment by developit** [comment]
That's even weirder. Maybe something from babel-polyfill interfering that is then skipped in Canary?

**Comment by slaskis** [comment]
Haha, spoke to soon, now it doesn't work anymore and I changed no code whatsoever. Just posted that comment and refreshed...

**Comment by developit** [comment]
Screams "race condition"!

**Comment by slaskis** [comment]
Haha, it sure does!

**Comment by developit** [comment]
This would make it work exactly like it did in `5.6`, though I do think it's worth figuring out what's causing the loop/race:

```js
import { options } from 'preact';

options.debounceRendering = fn => setImmediate(fn);
```

**Comment by slaskis** [comment]
Hmm. That still does not have the same effect as switching to 5.6.0. 

So, if the only change I do (from the version that works with 5.6) is upgrade to 5.7 and add that line it will re-mount over and over.

**Comment by slaskis** [comment]
I mean, there's most likely still a race condition somewhere in my actions, but just trying to narrow down the change that might make it visible

**Comment by developit** [comment]
Ok, the remounting thing seems to push it back into Preact's territory.  I do think that it's worth trying out the stuff since 5.7, but I also don't want to burn you out on debugging this.  Maybe for now revert to 5.6, and I'll try to get 5.8 out as quickly as I can, then we can revisit?

**Comment by slaskis** [comment]
There's something fishy for sure, but will all the poking around I've done today (updating babel, trying to change to bluebird for promises in an attempt to get more "debugging" info, adding logs everywhere, trying stepping around in dev tools etc.) just made me more confused. 

I'll gladly try out a beta version to see if it makes any difference, maybe it'll help me figure out my race condition too.

And regarding downgrading the main reason I started this was that I was not getting  `componentDidMount` triggered in an otherwise "dumb" component (it's quite far down the tree) but so far only on safari. 
But yeah, that should probably be another issue (will post one after some more testing).

And once again, thank you so much for being so helpful in this!

**Comment by developit** [comment]
Alright, I'll work on getting these commits organized for a proper (or beta) release. Definitely affects `componentDidMount`, so at least we know there will be a difference (can't say whether it will for sure fix the issue).  `react-redux` relies on mount events to connect and disconnect from the centralized store, so perhaps this is all just boiling down to that lifecycle.

**Comment by slaskis** [comment]
I'm not using `react-redux` in my app, I'm using [preact-socrates](https://github.com/matthewmueller/preact-socrates) to tie the stuff together if that might help. It's super simple and it doesn't seem to depend on any lifecycle events as far as I can see.

**Comment by developit** [comment]
Ah! Interesting, okay so that helps by removing a lot from the equation.  It might be useful to try injecting a `console.log()` [here](https://github.com/matthewmueller/preact-socrates/blob/master/lib/cycle.js#L38):

```js
  function dispatch () {
    console.log.apply(console, arguments);
    return store.apply(null, arguments)
  }
```

I think that should show the repeated actions being dispatched.

**Comment by slaskis** [comment]
Allright, good idea. Unfortunately it gave no more information that it's being dispatched from the `componentDidMount` and uses the dispatch function to get the current state inside the action (it's the same function in socrates).

**Comment by developit** [comment]
hmmmmm - if you wrap the dispatch in a setTimeout(), does that prevent it from looping?  Might be on to something here...

**Comment by developit** [comment]
@slaskis I just published [6.0.0 (beta)](https://github.com/developit/preact/releases/tag/6.0.0), which I'm hoping will fix any oddities on the preact side for you:
`npm i preact@beta`

**Comment by slaskis** [comment]
Yep, tried this now and no more re-rendering. Yay, well done sir! 👏

**Comment by developit** [comment]
Yayyyyy!
